### PR TITLE
Support custom data for Raygun Handler

### PR DESCRIPTION
**Pull request by Sazpaimon**
Second try. This time using the data from the context and extra arrays directly. This is similar to how other handlers work, such as the New Relic or Raven handler.

### CODE REVIEW

**Comment by wpillar** [comment]
@Sazpaimon hey, sorry for the delay. 

Would it make sense to include this logic in the formatter? At the moment `RayGunHandler::getDefaultFormatter()` is using a standard out of the box formatter, if we created a RayGun specific formatter (or perhaps something more generic if we can) then `getDefaultFormatter()` can return that and in the write methods we can use `$record['formatted']` which might be cleaner.

Thoughts?

**Comment by Sazpaimon** [comment]
The reason I didn't do it that way was because none of the other handlers I saw that implement something similar do it, mainly because they all had slightly different implementations where a shared formatter wouldn't make a lot of sense. I also didn't really see the benefit of a formatter just for Raygun, but I'm not opposed to it if that's what will get this PR merged.

**Comment by wpillar** [comment]
@Sazpaimon looks like the monolog project has formatters for specific handlers: https://github.com/Seldaek/monolog/tree/master/src/Monolog/Formatter

The `ChromePHPHandler` specifies its own formatter here: https://github.com/Seldaek/monolog/blob/master/src/Monolog/Handler/ChromePHPHandler.php#L94

I think there's sufficient precedent and I think it makes sense here with the amount of formatting logic we're adding.

**Comment by Sazpaimon** [comment]
Alright, I added a Raygun formatter and accompanying test. The formatter puts custom data/tags/timestamp into their own keys and removes them from the context and extra arrays since they aren't needed there anymore.

File: tests/unit/src/Graze/Monolog/Formatter/RaygunFormatterTest.php [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+            ),
+        );
+    }
+
+    /**
+     * @dataProvider logData
+     * @param array $input
+     * @param array $expected
+     */
+    public function testFormat(array $input, array $expected) {
```
Comment (wpillar) [inline tests/unit/src/Graze/Monolog/Formatter/RaygunFormatterTest.php]: This brace should be on the next line :)

**Comment by wpillar** [comment]
@Sazpaimon looking good, if you could fix that brace then I'll merge. I tried to fix it myself but realised I can't.

**Comment by Sazpaimon** [comment]
Done!

**Comment by wpillar** [comment]
@Sazpaimon thanks! and thanks for taking the time to contribute. I will tag it tomorrow once I decide whether it's a breaking change or not.

If you've got any thoughts on that, I'd love to hear them. I might change the new arguments you've added to be optional to preserve as much BC as possible.

**Comment by Sazpaimon** [comment]
I don't think there should be any BC breaks. If there's no custom data or tags, they just get passed as empty arrays, and the Raygun client will handle them as if they weren't passed in at all. Same with timestamp, which gets passes in as null if it isn't set.

**Comment by wpillar** [comment]
@Sazpaimon the potential BC breaks are the mandatory arguments added to `writeError` and `writeException`, they are protected so anyone extending them and calling the parent will experience an error simply by upgrading the version.

If we change those to optional arguments `array $tags = []` then that won't happen. Thoughts?

**Comment by Sazpaimon** [comment]
Ah, yeah that should be fine.

**Comment by wpillar** [comment]
Great, I'll take care of that tomorrow and tag it by the end of tomorrow. Sound good?

**Comment by Sazpaimon** [comment]
Great! I may have another PR coming as well to support sending Throwable in PHP 7 to Raygun, assuming the client supports it. I'll make sure to make that not BC breaking as well.

**Comment by wpillar** [comment]
@Sazpaimon cool, let's make sure it's still PHP5 compatible, I don't think we're quite ready to support a PHP 7 only version.

**Pull request merged**